sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala

```diff
@@ -147,7 +148,44 @@ case class SortPrefix(child: SortOrder) extends UnaryExpression {
       (!child.isAscending && child.nullOrdering == NullsLast)
   }
 
-  override def eval(input: InternalRow): Any = throw new UnsupportedOperationException
+  private lazy val calcPrefix: Any => Long = child.child.dataType match {
+    case BooleanType => (raw) =>
+      if (raw.asInstanceOf[Boolean]) 1 else 0
+    case _: IntegralType => (raw) =>
+      raw.asInstanceOf[java.lang.Number].longValue()
+    case DateType | TimestampType =>
+      _.asInstanceOf[java.lang.Number].longValue()
```

Comment:
Aren't the above two cases being the same thing?

Reply:
They are, we could try to specialize this a bit more by using `java.lang.Integer` and `java.lang.Long`.